File: leetcode/1654.到家的最少跳跃次数.cpp
class Solution {
public:
    int minimumJumps(vector<int>& forbidden, int a, int b, int x) {
        // 最远距离 bound = max(F + a + b, x + b)
        int F = *max_element(forbidden.begin(), forbidden.end()), bound = max(F + a + b, x + b);

        int ban[bound + 1];
        memset(ban, 0, sizeof(ban));
        for(int f : forbidden) {
            ban[f] = 1;
        }

        int dist[bound + 1][2]; // dist[i][0] - 上一次前跳, dist[i][1] - 上一次后跳
        memset(dist, 0x3f, sizeof(dist));
        dist[0][0] = 0;
        queue<pair<int,int>> q({{0, 0}});
        while(q.size()) {
            auto [i, pre] = q.front(); q.pop();
            if(i == x) {
                return dist[i][pre];
            }
            if(pre == 0 && i-b >= 0 && !ban[i-b] && dist[i][pre] + 1 < dist[i-b][1]) {
                dist[i-b][1] = dist[i][pre] + 1;
                q.emplace(i-b, 1);
            }
            if(i+a <= bound && !ban[i+a] && dist[i][pre] + 1 < dist[i+a][0]) {
                dist[i+a][0] = dist[i][pre] + 1;
                q.emplace(i+a, 0);
            }
        }

        return -1;
    }
};

File: leetcode/1654.到家的最少跳跃次数.py
"""
定义状态(idx, cnt), idx表示当前位置，cnt表示连续往后面跳了几次
- 注意设定上界, 要保证超过x还能跳回来
- 下界不能够跳成负数
- 采用vis记录状态
"""
class Solution:
    def minimumJumps(self, forbidden: List[int], a: int, b: int, target: int) -> int:
        up = max(max(forbidden) + a, target) + b
        forbidden = Counter(forbidden)
        q = deque([(0, 0)])
        vis = set()
        step = 0
        while q:
            m = len(q)
            for _ in range(m):
                x, cnt = q.popleft()
                if x == target:
                    return step
                dx = x + a
                if 0 <= dx <= up and (dx, 0) not in vis and dx not in forbidden:
                    vis.add((dx, 0))
                    q.append((dx, 0))
                dx = x - b
                if 0 <= dx <= up and (dx, cnt + 1) not in vis and dx not in forbidden:
                    if cnt < 1:
                        vis.add((dx, cnt + 1))
                        q.append((dx, cnt + 1))
            step += 1
        return -1